### PR TITLE
Implement SkeletonData

### PR DESCRIPTION
**Pull request by sanman00**
Implementation of `SkeletonData`.
- [x] Registration of field getters and setters
- [x] Accurately used the appropriate `AbstractData` implementation

Implementation of `ImmutableSkeletonData`
- [x] Accurately extended the appropriate `AbstractImmutableData` implementation
- [x] Accurately instantiating final instance fields with an `ImmutableValue` counter part for any value getters
- [x] If necessary, creating a new `ImmutableValue` for a value that should not be cached, or, using the existing caching utils.

Implementation of the `SkeletonTypeValueProcessor`(s):
- [x] Appropriately extended `AbstractSpongeValueProcessor`
- [x] Individual processors for each source type possible (one for `ItemStack`/`TileEntity`/`Entity` as necessary).

Registration:
- [x] Registered the `Key` correctly by `SKELETON_TYPE` in the `KeyRegistry`
- [x] Registered the `DataProcessor`  in `SpongeSerializationRegistry`
- [x] Registered the `ValueProcessor` in the `SpongeSerializationRegistry`

Test Plugin code:
```java
@Plugin(id = "skeleswapper", name = "SkeletonSwapper", version = "0.0.4_05")
public class Main {
    @Inject
    private Game game;

    public final CommandCallable printKeyCommand = CommandSpec
            .builder()
            .description(Texts.of("Prints the ", TextColors.YELLOW, "Keys.SKELETON_TYPE", TextColors.RESET, " key"))
            .executor((src, args) -> {
                System.out.println(Keys.SKELETON_TYPE);
                return CommandResult.success();
            })
            .build();

    public final CommandCallable unsupportedHolderCommand = CommandSpec
            .builder()
            .description(Texts.of("Tries to offer data to a ", TextColors.RED, "known", TextColors.RESET, " unsupported data holder"))
            .executor((src, args) -> {
                if (src instanceof Player) {
                    Player player = (Player) src;
                    DataTransactionResult result1 = player.offer(Keys.SKELETON_TYPE, SkeletonTypes.NORMAL);
                    player.sendMessage(
                            Texts.of("Results of offering the player skeleton data with a key: "),
                            Texts.of("Successful data: " + result1.getSuccessfulData()),
                            Texts.of("Replaced data: " + result1.getReplacedData()),
                            Texts.of("Rejected data: " + result1.getRejectedData()),
                            Texts.of("Type of result: " + result1.getType())
                    );

                    SkeletonData data = this.game.getRegistry().getManipulatorRegistry().getBuilder(SkeletonData.class).get().create();
                    System.out.println(data.type().get());
                    data = data.set(Keys.SKELETON_TYPE, SkeletonTypes.WITHER);
                    System.out.println(data.type().get());
                    DataTransactionResult result2 = player.offer(data);

                    player.sendMessage(
                            Texts.of("Results of offering the player skeleton data with a manipulator: "),
                            Texts.of("Successful data: " + result2.getSuccessfulData()),
                            Texts.of("Replaced data: " + result2.getReplacedData()),
                            Texts.of("Rejected data: " + result2.getRejectedData()),
                            Texts.of("Type of result: " + result2.getType())
                    );
                }

                else {
                    src.sendMessage(Texts.of(TextColors.RED, "This sender cannot hold data!"));
                }
                return CommandResult.success();
            })
            .build();
    @Listener
    public void onInit(GameInitializationEvent e) {
        e.getGame().getCommandDispatcher().register(this, this.printKeyCommand, "printkey");
        e.getGame().getCommandDispatcher().register(this, this.unsupportedHolderCommand, "unsupportedholder");
    }

    @Listener
    public void onEntityInteract(InteractEntityEvent e) {
        Entity targetEntity = e.getTargetEntity();
        Optional<SkeletonType> skeletonTypeOptional = targetEntity.get(Keys.SKELETON_TYPE);

        System.out.println("Value of the SkeletonType data in this entity: " + skeletonTypeOptional);
        SkeletonData data = this.game.getRegistry().getManipulatorRegistry().getBuilder(SkeletonData.class).get().create();

        if (skeletonTypeOptional.isPresent() && e.getCause().root().get() instanceof Player) {
            Player player = (Player) e.getCause().root().get();

            if (skeletonTypeOptional.get() == SkeletonTypes.NORMAL) {
                player.sendMessage(Texts.of("Changing ", TextColors.GRAY, "normal skeleton", TextColors.RESET, " to ", TextColors.BLACK, "wither skeleton"));
                System.out.println(targetEntity.get(Keys.SKELETON_TYPE));
                data = data.set(Keys.SKELETON_TYPE, SkeletonTypes.WITHER);
                System.out.println(data.toContainer().get(Keys.SKELETON_TYPE.getQuery()));
                targetEntity.offer(data);
            }

            else if (skeletonTypeOptional.get() == SkeletonTypes.WITHER) {
                player.sendMessage(Texts.of("Changing ", TextColors.BLACK, "wither skeleton", TextColors.RESET, " to ", TextColors.GRAY, "normal skeleton"));
                System.out.println(targetEntity.get(Keys.SKELETON_TYPE));
                data = data.set(Keys.SKELETON_TYPE, SkeletonTypes.NORMAL);
                System.out.println(data.toContainer().get(Keys.SKELETON_TYPE.getQuery()));
                targetEntity.offer(data);
            }
        }
    }
```

### CODE REVIEW

File: src/main/java/org/spongepowered/common/registry/SpongeGameRegistry.java [hunk 1]
```diff
@@ -708,6 +709,11 @@
             .put("west", (BigMushroomType) (Object) BlockHugeMushroom.EnumType.WEST)
             .build();
     
+    private final Map<String, SkeletonType> skeletonTypeMappings = new ImmutableMap.Builder<String, SkeletonType>()
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/registry/SpongeGameRegistry.java]: Please update your registry as the entire registry was changed. 

In short: Make a `SkeletonTypeRegistryModule` make the map, add the `RegisterCatalog` annotation, and add in `CommonModuleRegistry` `registry.registerModule(SkeletonType.class, new SkeletonTypeRegistryModule);`

File: src/main/java/org/spongepowered/common/mixin/core/entity/living/monster/MixinEntitySkeleton.java [hunk 2]
```diff
@@ -25,12 +25,56 @@
...
+import org.spongepowered.asm.mixin.Implements;
+import org.spongepowered.asm.mixin.Interface;
 import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
 
 @NonnullByDefault
 @Mixin(EntitySkeleton.class)
-public abstract class MixinEntitySkeleton extends MixinEntityMob implements Skeleton {
+@Implements(@Interface(iface = Skeleton.class, prefix = "skeleton$"))
+public abstract class MixinEntitySkeleton extends MixinEntityMob {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/entity/living/monster/MixinEntitySkeleton.java]: You should be able to actually implement `Skeleton`.

**Comment by sanman00** [comment]
@gabizou Is there anything else I need to do with this pr?

**Comment by sanman00** [comment]
Closing in favour of #276.

**Pull request closed without merging**